src/uu/cp/src/platform/linux.rs

```diff
@@ -132,6 +135,43 @@ where
     Ok(num_bytes_copied)
 }
 
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn check_dest_is_fifo(dest: &Path) -> bool {
+    //If our destination file exists and its a fifo , we do a standard copy .
+    let file_type = fs::metadata(dest);
+    match file_type {
+        Ok(f) => f.file_type().is_fifo(),
+
+        _ => false,
+    }
+}
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+fn check_sparse_detection(source: &Path) -> Result<bool, std::io::Error> {
+    let mut src_file = File::open(source)?;
+
+    let size = src_file.metadata()?.size();
+    let blocks = src_file.metadata()?.blocks();
+    let blksize = src_file.metadata()?.blksize();
+
+    let mut data_flag = false;
```

Comment:
I think it can be simplified with something like:

    if blocks < size / 512 || blocks == 0 {
```suggestion
    let mut data_flag = false;
```
        return Ok(true);
    }

    let mut buf = vec![0; blksize as usize];
    src_file.read(&mut buf)?;
    if buf.iter().any(|&x| x != 0) {
        return Ok(false);
    }

    Ok(true)

Reply:
or something like this :)

(early exit)

Reply:
``` 
if blocks == 0 {
        src_file.read(&mut buf)?;
        if buf.iter().any(|&x| x != 0x0) {
            return Ok(false);
        } else {
            return Ok(true);
        }
    }

    if blocks < size / 512 {
        return Ok(true);
    }
    Ok(false)
```
Would this be okay ? So basically we need to read the file only if the file consists of zero blocks as virtual files also show zero blocks in its metadata . If a zero block file has data then its not sparse , if it has no data then its sparse.
Then the other non zero block files can be categorized by the formula . I will write a comment there since  the code isn't self explanatory .

Reply:
sounds good